### ISSUE TITLE
Use ts.forEachChild more

### ISSUE THREAD

**Issue by dsherret** [dsherret/ts-simple-ast#281]
I believe performance improvements could be gained by internally updating the library to use `ts.forEachChild` more.

**Comment by cancerberoSgx** [comment]
regarding this, I think you are right - but take in account the semantics of getChildren and forEachChildren are different - even the returned structure is different, although you probably already know this. 

Regarding performance, just in case read https://github.com/Microsoft/TypeScript/issues/23917#issuecomment-388176940
What that guy say to me, regarding performance, is to pass  the sourceFile to Node methods that optionally accept it in case you already have a reference to it. If you don't pass it the implementation has to climb up the tree to the sourceFile, each time! methods are getStart(), getWidth, getChildren() and others. I think they want to make that sourceFile parameter not optionally any more but now its too late. My two cents.

**Comment by dsherret** [comment]
@cancerberoSgx yes, there's an option shown on https://ts-ast-viewer.com ("Tree mode") that shows the difference between the two.

The implementations in this library of methods that require a source file to be passed to them in the compiler API (`.getStart()`, `.getWidth()`, etc.) all pass the source file for you. For example:

```ts
/**
 * Gets the width of the node (length without trivia).
 */
getWidth() {
    return this.compilerNode.getWidth(this.sourceFile.compilerNode);
}
```

One point about `.getChildren()` is that calling it requires additional parsing (unless the compiler has previously cached the result of the parse). That's why a performance improvement could be gained by avoiding it if necessary, but I'm unsure how much of a performance improvement. It might not be much so I'll have to test it to see.

**Comment by cancerberoSgx** [comment]
BTW I'm playing / learning to about vscode extensions and I'm making a AST outline view based on vscode TreeView and ts-simple-ast : https://marketplace.visualstudio.com/items?itemName=cancerberosgx.vscode-typescript-ast-outline

Have already working some stuff like remove node, rename node, and toggle AST mode. This last one clearly shows the difference between forEachChild and getChildren BTW. 

I'm testing the extension with  very large projects like this one, vscode and typescript and its behaving pretty acceptable although is not lightweight.  - Just two days working on it so it's not ready at all. The objective is to have a outline like Java/Eclipse c# etc where you can drag&drop and perform refactors visually . Now I just know that vscode has its own outline implementation but it's very simple compared to what I want. Thought you might find it interesting. Thanks

**Comment by cancerberoSgx** [comment]
Yes that previous issue link I shared with you is about that - it happened to me that getChildren was failing when climbing up the AST to find the sourcefile . The solution/workaround was or pass the sourcefile on the call or implement compilerHost.getSourceFile as shown in second comment. Thanks

**Comment by dsherret** [comment]
@cancerberoSgx cool! That seems like a neat project. I hope it turns out well! 😃

**Comment by dsherret** [comment]
Implemented in #307. Will update code as I see it.
